processor/marshalprocessor/processor.go

```diff
@@ -0,0 +1,136 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package marshalprocessor
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+type marshalProcessor struct {
+	logger          *zap.Logger
+	marshalTo       string
+	kvSeparator     rune
+	kvPairSeparator rune
+}
+
+func newMarshalProcessor(logger *zap.Logger, cfg *Config) *marshalProcessor {
+	return &marshalProcessor{
+		logger:          logger,
+		marshalTo:       cfg.MarshalTo,
+		kvSeparator:     cfg.KVSeparator,
+		kvPairSeparator: cfg.KVPairSeparator,
+	}
+}
+
+func (mp *marshalProcessor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resourceLog := ld.ResourceLogs().At(i)
+		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
+			scopeLog := resourceLog.ScopeLogs().At(j)
+			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
+				logRecord := scopeLog.LogRecords().At(k)
+				logBody := logRecord.Body()
+				// If body is not a map, skip that log
+				if logBody.Type() != pcommon.ValueTypeMap {
+					mp.logger.Warn("Non map body not supported", zap.Any("body", logBody))
+					continue
+				}
+				switch strings.ToUpper(mp.marshalTo) {
+				case "JSON":
+					jsonBody := logBody.AsString()
+					logBody.SetStr(jsonBody)
+				case "XML":
+					return ld, fmt.Errorf("XML not yet supported")
+				case "KV":
+					kvBody := mp.convertToKV(logBody.Map())
+					logBody.SetStr(kvBody)
+				default:
+					return ld, fmt.Errorf("Unrecognized format to marshal to: %s", mp.marshalTo)
+				}
+			}
+		}
+	}
+
+	return ld, nil
+}
+
+func (mp *marshalProcessor) convertToKV(logBody pcommon.Map) string {
+	var kvStrings []string
+	var kvPairSeparator string
+	var kvSeparator string
+	if mp.kvPairSeparator == 0 {
+		kvPairSeparator = " "
+	} else {
+		kvPairSeparator = string(mp.kvPairSeparator)
+	}
+	if mp.kvSeparator == 0 {
+		kvSeparator = "="
+	} else {
+		kvSeparator = string(mp.kvSeparator)
+	}
+
+	for k, v := range logBody.AsRaw() {
+		k = strings.ReplaceAll(k, "\"", "\\\"")
+		if strings.Contains(k, kvPairSeparator) || strings.Contains(k, kvSeparator) {
+			k = "\"" + k + "\""
+		}
+
+		if valMap, ok := v.(map[string]interface{}); ok {
+			v = convertMapToString(valMap)
+		}
```

Comment:
Why do we necessarily need a separate function `convertMapToString` here? `convertToKV` already converts a map to KV format. 

Maybe instead of using logBody.AsRaw we could iterate over the pcommon values using `https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pcommon#Map.Range` and recursively call `convertToKV` instead of having a separate function that seems very similar.

Reply:
Values in nested maps have to be escaped differently. But doing it recursively still makes sense.